File: _CodeTopics/LeetCode/201-400/000203/000203_algo2.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """

        # 递归法实现：假定后面的都已经删好了。

        if not head:
            return None
        elif head.val != val:
            head.next = self.removeElements(head.next, val)
        else:
            return self.removeElements(head.next, val)
        return head
        
"""
https://leetcode-cn.com/submissions/detail/184190237/

66 / 66 个通过测试用例
状态：通过
执行用时: 52 ms
内存消耗: 27.4 MB

执行用时：52 ms, 在所有 Python 提交中击败了82.48%的用户
内存消耗：27.4 MB, 在所有 Python 提交中击败了5.10%的用户
"""

File: _CodeTopics/LeetCode/201-400/000203/tran/000203_algo2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {

        if (head == nullptr) {
            return nullptr;
        } else if (head->val != val) {
            head->next = this->removeElements(head->next, val);
        } else {
            // 加不加 this 都行。
            return removeElements(head->next, val);
        }
        return head;
    }
};

/*
https://leetcode-cn.com/submissions/detail/184195463/

66 / 66 个通过测试用例
状态：通过
执行用时: 16 ms
内存消耗: 15.1 MB

执行用时：16 ms, 在所有 C++ 提交中击败了99.70%的用户
内存消耗：15.1 MB, 在所有 C++ 提交中击败了5.03%的用户
*/
